src/huntsman/pocs/states/huntsman/coarse_focusing.py

```diff
@@ -27,24 +24,48 @@ def on_enter(event_data):
 
     # Convert altaz coordinates to radec.
     coarse_focus_coords = altaz_to_radec(alt=coarse_focus_alt, az=coarse_focus_az,
-                                         location=earth_location, obstime=current_time())
+                                         location=pocs.observatory.earth_location,
+                                         obstime=current_time())
 
     pocs.say(f'Coarse focus coordinates: {coarse_focus_coords}')
 
     # Slew to coordinates for coarse focusing.
     pocs.say("I'm slewing over to the coordinates for coarse focus.")
     pocs.observatory.mount.slew_to_coordinates(coarse_focus_coords)
-
     pocs.say("I'm at the coordinates for coarse focus.")
 
-    # Do the autofocusing
-    pocs.say("Coarse focusing all cameras.")
+    # Do the autofocusing (non-blocking).
+    pocs.say("Starting coarse focus on all cameras.")
     autofocus_events = pocs.observatory.autofocus_cameras(coarse=True)
+
+    timer = CountdownTimer(coarse_focus_timeout)
     pocs.logger.debug("Waiting for coarse focus to finish.")
-    wait_for_events(list(autofocus_events.values()), timeout=coarse_focus_timeout)
 
-    # Morning and not dark enough for observing...
-    if pocs.observatory.past_midnight and not pocs.is_dark(horizon='observe'):
-        pocs.next_state = 'twilight_flat_fielding'
-    else:
-        pocs.next_state = 'scheduling'
+    while True:
+        # Check if all cameras are finished and proceed if so.
+        if all([e.is_set() for e in autofocus_events.values()]):
+            break
+
+        # If overall timer has expired, check which cameras have not finished
+        # and remove them from the observatory.
+        if timer.expired():
+            pocs.logger.info(f'Autofocus timer expired, checking which cameras were successful')
+            for cam_name, focus_event in autofocus_events.items():
+                if not focus_event.is_set():
+                    pocs.logger.warning(f'Incomplete focus event on {cam_name}, removing')
+                    # TODO figure out why and also mark them for restart somehow.
+                    # See https://github.com/AstroHuntsman/huntsman-pocs/issues/304
+                    pocs.observatory.remove_camera(cam_name)
```

Comment:
I've seen problems before where changing the size of dictionaries while iterating over them can cause errors (e.g., https://stackoverflow.com/questions/11941817/how-to-avoid-runtimeerror-dictionary-changed-size-during-iteration-error).

I'm not sure whether we will have that problem here? If this is already tested in `POCS` then it should be fine.

Reply:
This is iterating over a a dictionary of just camera names and `threading.Event` objects, so I think we should be okay.